### PR TITLE
Add ramping limit options for conventional powerplants in prepare_sector_network.

### LINKED ISSUE #20: Include ramping limit for nuclear power plants
Following steps are necessary:
- [ ] Include ramping limit paramter to nuclear power plant generators
- [ ] Research online available ramping limits and take conservative approach for value

### PR DESCRIPTION
**Pull request by virio-andreyana**
Closes #20 

## Checklist

- [x] I tested my contribution locally and it works as intended.
- [ ] Code and workflow changes are sufficiently documented.
- [ ] Changed dependencies are added to `envs/environment.yaml`.
- [ ] Changes in configuration options are added in `config/config.default.yaml`.
- [ ] Changes in configuration options are documented in `doc/configtables/*.csv`.
- [ ] Sources of newly added data are documented in `doc/data_sources.rst`.
- [ ] A release note `doc/release_notes.rst` is added.

### CODE REVIEW

**Comment by daniel-rdt** [comment]
Looks good from my side. Thanks @virio-andreyana! Some notes:
- Let's add this to the default config as well to make it compatible with upstreaming the changes at some point
- This would include adding the changes to the doc files and adding the source for the default assumptions
- The ramp limit should also be added following the same logic in the 'add_existing_baseyear script' here:
https://github.com/open-energy-transition/form-energy-storage/blob/10e8d8b94e75b3f825f947b21da31a0d11ed442d/scripts/add_existing_baseyear.py#L320-L337
- I am considering if we should include the `p_min_pu` for nuclear power plants as well here and call this PR something like `improvement of nuclear power plant modeling`

**Comment by virio-andreyana** [comment]
I just found a subtle way to include this without adding any new code within the script.
so at the end of `prepare_network.py` and `prepare_sector_network.py` there is this function `maybe_adjust_costs_and_potentials()` which leads to `modify_attribute()` which then can update any components in a PyPSA based on the config adjustment. So for example:

```yaml
# docs in https://pypsa-eur.readthedocs.io/en/latest/configuration.html#adjustments
adjustment:
  electricity:
    absolute:
      Generator:
        nuclear:
          ramp_limit_up: 0.5
          ramp_limit_up: 0.5
          p_min_pu: 0.5
  sector:
    absolute:
      Link:
        nuclear:
          ramp_limit_up: 0.5
          ramp_limit_up: 0.5
          p_min_pu: 0.5
```

automatically add what we want without changing any of the scripts. What do you think @daniel-rdt ? If you agree, I'll clean everything up and just put this in config.default.yaml instead.

**Comment by virio-andreyana** [comment]
Ready for review @tgilon @daniel-rdt. I've neglected my previous suggestion and continue with this one.

**Comment by daniel-rdt** [comment]
I need some more time to review this thoroughly. Thus, I will not merge this into the dev branch yet and will start the updated run without the ramping and p_min_pu. We can decide after and include it then.

File: config/config.form.yaml [hunk 1]
```diff
@@ -52,7 +52,7 @@ snapshots:
 # docs in https://pypsa-eur.readthedocs.io/en/latest/configuration.html#enable
 enable:
   retrieve: auto
-  retrieve_databundle: true
+  retrieve_databundle: false
```
Comment (tgilon) [inline config/config.form.yaml]: Not sure this has to be changed

File: scripts/prepare_sector_network.py [hunk 2]
```diff
@@ -1235,6 +1235,9 @@ def add_generation(n, costs, existing_capacities=0, existing_efficiencies=None):
             ),
             efficiency2=costs.at[carrier, "CO2 intensity"],
             lifetime=costs.at[generator, "lifetime"],
+            p_min_pu=cf_conventional[generator]["p_min_pu"] if "p_min_pu" in cf_conventional.get(generator,{}).keys() else 0,
```
Comment (tgilon) [inline scripts/prepare_sector_network.py]: We can simplify the expression

```suggestion
            p_min_pu=cf_conventional.get(generator,{}).get("p_min_pu", 0),
```

**Review by tgilon — COMMENTED** [review]
I like the idea of having all the nuclear tweaks at the same place in `conventional`.

**Comment by virio-andreyana** [comment]
add @tgilon suggestion

**Comment by tgilon** [comment]
" France, a nuclear power plant is able to ramp up or down between 100% and 20% of nominal power in half an hour, and again after at least two hours, twice a day."

source : https://edf.hal.science/hal-01977209v1/document

File: config/config.default.yaml [hunk 3]
```diff
@@ -279,6 +279,9 @@ conventional:
   dynamic_fuel_price: false
   nuclear:
     p_max_pu: "data/nuclear_p_max_pu.csv" # float of file name
+    p_min_pu: 0.5
```
Comment (daniel-rdt) [inline config/config.default.yaml]:
```suggestion
    p_min_pu: 0.2
```

File: config/config.default.yaml [hunk 4]
```diff
@@ -279,6 +279,9 @@ conventional:
...
+    ramp_limit_up: 0.5
```
Comment (daniel-rdt) [inline config/config.default.yaml]:
```suggestion
```

File: config/config.default.yaml [hunk 5]
```diff
@@ -279,6 +279,9 @@ conventional:
...
+    ramp_limit_down: 0.5
```
Comment (daniel-rdt) [inline config/config.default.yaml]:
```suggestion
```

File: config/config.form.yaml [hunk 6]
```diff
@@ -100,6 +100,16 @@ renewable:
   hydro:
     PHS_max_hours: 10
 
+# docs in https://pypsa-eur.readthedocs.io/en/latest/configuration.html#conventional
+conventional:
+  unit_commitment: false
+  dynamic_fuel_price: false
+  nuclear:
+    p_max_pu: "data/nuclear_p_max_pu.csv" # float of file name
+    p_min_pu: 0.5
```
Comment (daniel-rdt) [inline config/config.form.yaml]:
```suggestion
    p_min_pu: 0.2
```

File: config/config.form.yaml [hunk 7]
```diff
@@ -100,6 +100,16 @@ renewable:
...
+    ramp_limit_up: 0.5
```
Comment (daniel-rdt) [inline config/config.form.yaml]:
```suggestion
```

File: config/config.form.yaml [hunk 8]
```diff
@@ -100,6 +100,16 @@ renewable:
...
+    ramp_limit_down: 0.5
```
Comment (daniel-rdt) [inline config/config.form.yaml]:
```suggestion
```

**Review by daniel-rdt — COMMENTED** [review]
Updated `p_min_pu`to 20% and removed ramping limits based on EDF source as quoted by @tgilon:
> " France, a nuclear power plant is able to ramp up or down between 100% and 20% of nominal power in half an hour, and again after at least two hours, twice a day."

source : https://edf.hal.science/hal-01977209v1/document

File: scripts/add_existing_baseyear.py [hunk 9]
```diff
@@ -334,7 +336,10 @@ def add_power_capacities_installed_before_baseyear(n, grouping_years, costs, bas
                         efficiency2=costs.at[carrier[generator], "CO2 intensity"],
                         build_year=grouping_year,
                         lifetime=lifetime_assets.loc[new_capacity.index],
-                    )
+                        p_min_pu=cf_conventional.get(generator,{}).get("p_min_pu",0),
+                        ramp_limit_up=cf_conventional.get(generator,{}).get("ramp_limit_up",np.nan),
+                        ramp_limit_down=cf_conventional.get(generator,{}).get("ramp_limit_down",np.nan),
+            )
```
Comment (tgilon) [inline scripts/add_existing_baseyear.py:342]: FYI @daniel-rdt Spacing before `)` has been changed

**Pull request merged**